### PR TITLE
router: switch to from net/http/httputil to net/http

### LINKED ISSUE #61: Persistent backend connections in router
Due to flynn-archive/strowger#26, we changed the router to close the HTTP backend connection after each request in flynn-archive/strowger#27. This is not optimal, and we need a long-term solution that allows us to have a 1:1 mapping of client connections to backend connections. The main thing that needs to be solved is detection of connections closed by the backend before we send the next request.

**Relevant IRC logs**

- https://botbot.me/freenode/flynn/2014-06-19/?msg=16568761&page=1
- https://botbot.me/freenode/flynn/2014-06-20/?msg=16619353&page=1
- https://botbot.me/freenode/flynn/2014-06-20/?msg=16657413&page=1

### LINKED ISSUE #64: HTTP/1.0 support in router
Currently the router only supports HTTP/1.1, but clients exist that use HTTP/1.0 (mostly telnet, probably). See flynn-archive/flynn-demo#31 for discussion.

### PR DESCRIPTION
**Pull request by bgentry**
This PR switches Flynn's router away from using the somewhat-deprecated `httputil.ClientConn` and `httputil.ServerConn` to using `net/http` HTTP parsers and handlers.

Because there are subtle differences between handling protocol upgrades (i.e. for websockets) and regular request proxying, there's a lot of duplicated logic between the functions that handle these paths. I think we'll be able to clean that up a lot, but would rather this doesn't spend too much more time in a side branch :)

We pretty quickly had to move away from `httputil.ReverseProxy` as it makes it basically impossible to do some of the stuff we want to do with trying conns to multiple backends, and it also wouldn't work for the websocket upgrade. In doing so, we've currently lost some of the nice parts of that package like periodic flushing of buffers (see [ReverseProxy.FlushInterval](http://golang.org/pkg/net/http/httputil/#ReverseProxy)). I've left a couple of TODOs for that in the code and intend to bring it in in a future PR. I also think that we can better package up some of the stuff we've done into a nice encapsulation similar to `httputil.ReverseProxy`, but again, didn't want to branch off for too long.

Finally, the way we pass down the sticky session setting from the route to the service is pretty ugly (using a temporary internal header). I didn't see an obvious way to access this bool (`r.Sticky`) from the part of the code where I needed to use it, so this was a compromise for expediency. I suspect we can find a much better way to do this, either by refactoring to make the setting available where we need it or by storing that setting inside of [a request context](https://blog.golang.org/context).

Fixes #61. Fixes #64. Makes some progress on #67 by setting some sane defaults for TCP timeouts.

### CODE REVIEW

**Comment by bgentry** [comment]
one fail so far in the Flynn CI:

```
FAIL: test_git_deploy.go:72: GitDeploySuite.TestRubyBuildpack

++ 18:47:44.769 /usr/bin/git clone https://github.com/flynn-examples/ruby-flynn-example /tmp/check-8674665223082153551/10/repo
Cloning into '/tmp/check-8674665223082153551/10/repo'...

++ 18:47:45.999 /home/ubuntu/go/src/github.com/flynn/flynn/test/../cli/flynn-cli create ruby-flynn-example
Created ruby-flynn-example

++ 18:47:46.581 /usr/bin/git push flynn master
remote: [1G[K-----> Building ruby-flynn-example...        
remote: Build failed: cluster: no servers found        
To ssh://git@flynn-1bd56847017dad2c.local:2222/ruby-flynn-example.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'ssh://git@flynn-1bd56847017dad2c.local:2222/ruby-flynn-example.git'

test_git_deploy.go:73:
    s.runBuildpackTest(t, "ruby-flynn-example", nil)
test_git_deploy.go:106:
    t.Assert(push, Succeeds)
... result *main.CmdResult = &main.CmdResult{Cmd:[]string{"git", "push", "flynn", "master"}, Output:"remote: \x1b[1G\x1b[K-----> Building ruby-flynn-example...        \nremote: Build failed: cluster: no servers found        \nTo ssh://git@flynn-1bd56847017dad2c.local:2222/ruby-flynn-example.git\n ! [remote rejected] master -> master (pre-receive hook declined)\nerror: failed to push some refs to 'ssh://git@flynn-1bd56847017dad2c.local:2222/ruby-flynn-example.git'\n", Err:(*exec.ExitError)(0xc2083f8400)}
```

File: router/handlers_test.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+var nopHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {})
+
+func TestFwdProtoHandler(t *testing.T) {
```
Comment (benburkert) [inline router/handlers_test.go]: are we :ok_hand: with mixing the use of `testing.T` & `check.C` in the router?

Comment (bgentry) [inline router/handlers_test.go]: yeah not sure. We had these tests already from one of our own project so I just imported them as-is (with additions for `X-Forwarded-For`). Let me know if you want to switch this to using gocheck instead.

File: router/http.go [hunk 2]
```diff
@@ -275,62 +303,28 @@ func (s *HTTPListener) findRouteForHost(host string) *httpRoute {
...
-		StatusCode:    code,
-		ProtoMajor:    1,
-		ProtoMinor:    0,
-		Request:       req,
-		Body:          ioutil.NopCloser(bytes.NewBufferString(msg)),
-		ContentLength: int64(len(msg)),
-	}
-	sc.Write(req, resp)
+func fail(w http.ResponseWriter, code int, msg string) {
+	w.Header().Set("Content-Length", strconv.Itoa(len(msg)))
```
Comment (benburkert) [inline router/http.go]: Is this to force a fixed length response? Do we want to `Connection: close` on an error?

File: router/http.go [hunk 3]
```diff
@@ -422,109 +376,269 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
...
-		if prior, ok := req.Header["X-Forwarded-For"]; ok {
-			clientIP = strings.Join(prior, ", ") + ", " + clientIP
+	isSticky := false
+	stickyAddr := ""
+	if req.Header.Get(hdrUseStickySessions) == "true" {
+		// TODO(bgentry): switch to better way to check sticky setting
+		isSticky = true
+		req.Header.Del(hdrUseStickySessions)
+		if stickyAddr = s.stickyCookieAddr(req); stickyAddr != "" {
+			sortStringFirst(addrs, stickyAddr)
```
Comment (benburkert) [inline router/http.go]: :+1: I like this approach.

File: router/http.go (hunk 2 above)
Comment (bgentry) [inline router/http.go]: It does force a fixed-length (non-chunked) response. As far as whether we want to `Connection: close`, after an error, that's generally a no, but there may be cases (specifically most `fail()` calls inside the TCP proxying path) where we do want to close the conn.

We can either add a param to `fail()` to indicate that, or just call `w.Header().Set("Connection", "close")` in each specific place where we want to do that. Extra param sounds better to me.

Comment (bgentry) [inline router/http.go]: also, that means more subtle differences between the two paths :(

File: router/http_test.go [hunk 4]
```diff
@@ -401,6 +409,65 @@ func (s *S) TestStickyHTTPRoute(c *C) {
...
+	defer l.Close()
+
+	addStickyHTTPRoute(c, l)
+
+	discoverdRegisterHTTP(c, l, srv1.Listener.Addr().String())
+	defer discoverd.UnregisterAll()
+
+	// triggers the TCP-mode conn upgrade path, but does not hit the websocket
+	// handler and so the upgrade is not successful.
+	assertGetCookieUpgradeClosed := func(c *C, url, host, expected string, cookie *http.Cookie) *http.Cookie {
```
Comment (benburkert) [inline router/http_test.go]: this is really funky. It could probably be rewritten as table driven tests.

Comment (bgentry) [inline router/http_test.go]: can it really work as a table? The ordering of events matters here.

File: router/http_test.go [hunk 5]
```diff
@@ -401,6 +409,65 @@ func (s *S) TestStickyHTTPRoute(c *C) {
 	}
 }
 
+func (s *S) TestStickyHTTPRouteWebsocket(c *C) {
+	wshandler := func(w http.ResponseWriter, req *http.Request) {
+		w.Write([]byte("ok " + strings.TrimPrefix(req.URL.Path, "/")))
```
Comment (benburkert) [inline router/http_test.go]: because the request path is used for the response body, it's possible for a request to `srv2` with path `/1` to respond with `"ok 1"`.

**Comment by benburkert** [comment]
I think this is a big step forward, glad to see this coming together quickly. My two concerns are that the `TestStickyHTTPRouteWebsocket` test doesn't actually perform the websocket handshake, and that it's hard to follow along with how the vendored code from `reverseproxy.go` has been patched. For the second concern, I'm gonna look into refactoring the proxy code into a `router/proxy` package.

File: router/http_test.go (hunk 5 above)
Comment (bgentry) [inline router/http_test.go]: ugh, I knew this ended up being too simple after I cleaned it up. Good catch.

**Comment by benburkert** [comment]
> Finally, the way we pass down the sticky session setting from the route to the service is pretty ugly (using a temporary internal header). I didn't see an obvious way to access this bool (r.Sticky) from the part of the code where I needed to use it, so this was a compromise for expediency. I suspect we can find a much better way to do this, either by refactoring to make the setting available where we need it or by storing that setting inside of a request context.

I agree 100%, and I suspect that adding a context to the request path will also be useful for addressing #67.

File: router/handlers_test.go (hunk 1 above)
Comment (titanous) [inline router/handlers_test.go]: I'd prefer changing this to gocheck for consistency.

File: router/http.go [hunk 6]
```diff
@@ -275,62 +303,28 @@ func (s *HTTPListener) findRouteForHost(host string) *httpRoute {
...
-		}
-
-		req.RemoteAddr = conn.RemoteAddr().String()
-		if r.service.handle(req, sc, isTLS, r.Sticky) {
-			return
-		}
+	// TODO(bgentry): find a better way to access this setting in the service
+	// where it's needed.
+	if r.Sticky {
+		req.Header.Set(hdrUseStickySessions, "true")
```
Comment (titanous) [inline router/http.go]: +1 for using a context here. In the mean time, always set this header so that it can't be injected.

File: router/http.go [hunk 7]
```diff
@@ -422,109 +376,269 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
...
 	}
-	if tls {
-		req.Header.Set("X-Forwarded-Proto", "https")
-	} else {
-		req.Header.Set("X-Forwarded-Proto", "http")
-	}
-	// TODO: Set X-Forwarded-Port
+
+	// Most of this is borrowed from httputil.ReverseProxy
+	outreq := new(http.Request)
```
Comment (titanous) [inline router/http.go]: We've standardized on not using `new` to initialize structs, use `&http.Request{}`.

File: router/http.go (hunk 6 above)
Comment (bgentry) [inline router/http.go]: *facepalm* great catch.

**Comment by titanous** [comment]
> one fail so far in the Flynn CI:

This is an unrelated failure.

File: router/http.go (hunk 2 above)
Comment (bgentry) [inline router/http.go]: opted for a `failAndClose()` that just calls `fail()` after setting the `Connection: close` header. See cfff35ce2f3123fc6703d9bfab9d7847615915d3

File: router/http.go (hunk 6 above)
Comment (bgentry) [inline router/http.go]: fixed in fc0efc7622c2ac8a609897d557208f85aa34b029

File: router/http.go [hunk 8]
```diff
@@ -422,109 +383,271 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
...
+	}
+
+	if res.StatusCode != 101 {
+		return
+	}
+	res.Body.Close() // close this now since we've copied everything
+
+	hj, ok := w.(http.Hijacker)
+	if !ok {
+		log.Println("not a hijacker")
```
Comment (titanous) [inline router/http.go]: Is it reasonable to panic here? There should be no situation where we hit this code path with the current code.

File: router/handlers_test.go (hunk 1 above)
Comment (bgentry) [inline router/handlers_test.go]: Done. c6fa8810ca8be812adca3659bb9abe25b0ea1a91

File: router/http.go (hunk 8 above)
Comment (bgentry) [inline router/http.go]: yeah, I think so on this particular error. What do you think about the line down below (logged with `hijack failed:`)?

Comment (titanous) [inline router/http.go]: Currently we would never hit that path either: https://github.com/golang/go/blob/4b96409aacfb569ac2564240d8aadd668b7c25fa/src/net/http/server.go#L142-L147

But there's no guarantee that other more dynamic errors won't be added in the future.

Comment (bgentry) [inline router/http.go]: yeah, ok, I vote to keep that other scenario as a non-panic.

File: router/http_test.go (hunk 4 above)
Comment (bgentry) [inline router/http_test.go]: FYI @benburkert has reorganized this test case so that it's much clearer what's going on: https://github.com/bgentry/flynn/pull/1

Will bring that in after the last couple of tweaks.

File: router/http.go [hunk 9]
```diff
@@ -422,109 +383,271 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
...
+		outreq.URL.Host = backend
+		res, err = transport.RoundTrip(outreq)
+		if err != nil {
+			if _, ok := err.(*dialErr); ok {
+				// retry, maybe log a message about it
+				continue
 			}
-			return true
+			log.Println("http: proxy error:", err)
+			fail(w, 503, http.StatusText(503))
```
Comment (titanous) [inline router/http.go]: Seems like `fail` could be modified to do the status text lookup itself? `fail(w, 503)`

Comment (bgentry) [inline router/http.go]: yeah, as long as you're ok with always using the default text for that status code. Looks like that's pretty much what's already happening in the other places that were using it.

Comment (titanous) [inline router/http.go]: Yeah, I expect we'll have nice error pages at some point, but for now it looks like we're just using variants of the status text.

File: router/http.go (hunk 8 above)
Comment (bgentry) [inline router/http.go]: Fixed in 2fd76aeed2ed65ca08316cbe4c194b4e28c88594

File: router/http.go (hunk 7 above)
Comment (bgentry) [inline router/http.go]: Fixed in 7d7dd9d8996968d862d786722b78f2d38b85edee

File: router/http_test.go (hunk 4 above)
Comment (bgentry) [inline router/http_test.go]: Cherry-picked that PR as 89e66509b7a59097d4e443c5fc5baa161bf08a8b

File: router/http.go (hunk 9 above)
Comment (bgentry) [inline router/http.go]: Fixed in 4245d38.

**Comment by bgentry** [comment]
I think I've resolved everything mentioned so far. :beers:

File: router/http.go [hunk 10]
```diff
@@ -275,62 +303,36 @@ func (s *HTTPListener) findRouteForHost(host string) *httpRoute {
...
 
-var errMissingTLS = errors.New("router: route not found or TLS not configured")
-
-func (s *HTTPListener) handle(conn net.Conn, isTLS bool) {
-	defer conn.Close() // TODO(benburkert): for TLS, check that closing this conn sends the TLS "close notify"
+func fail(w http.ResponseWriter, code int) {
+	msg := []byte(http.StatusText(code))
+	w.Header().Set("Content-Length", strconv.Itoa(len(msg)))
+	w.WriteHeader(code)
+	w.Write(msg)
```
Comment (titanous) [inline router/http.go]: Add a trailing newline so that when we curl it looks better?

File: router/http.go [hunk 11]
```diff
@@ -422,109 +384,269 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
...
+		return
+	}
+
+	if res.StatusCode != 101 {
+		return
+	}
+	res.Body.Close() // close this now since we've copied everything
+
+	hj, ok := w.(http.Hijacker)
+	if !ok {
```
Comment (titanous) [inline router/http.go]: Can just type assert without checking: `w.(http.Hijacker).Hijack()` which will have an even better error message in the panic.

**Comment by titanous** [comment]
This looks great! Please squash any minor fixup commits, after the last two comments are addressed we should be good to merge. :dancer:

**Comment by bgentry** [comment]
I think I've squashed all the ones that make sense. Lmk if you see any others worth squashing down.

**Comment by titanous** [comment]
LGTM, will merge when CI looks good.

**Comment by titanous** [comment]
:tada:

**Comment by bgentry** [comment]
:v:
